cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,160 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
                      programTerminator, ownerAdmin.getOwner(appId), appRequest.canUpdateSchedules());
   }
 
+  /**
+   * Upgrades an existing application by upgrading appliation artifact versions and plugin artifact versions.
+   *
+   * @param appId the id of the application to upgrade.
+   * @param programTerminator a program terminator that will stop programs that are removed when updating an app.
+   *                          For example, if an update removes a flow, the terminator defines how to stop that flow.
+   * @return Details of upgrade for the application like failure details with reason.
+   * @return
+   * @throws Exception
+   */
+  public ApplicationUpdateDetails upgradeApplication(ApplicationId appId, ProgramTerminator programTerminator)
+      throws Exception {
+    // Check if the current user has admin privileges on it before updating.
+    authorizationEnforcer.enforce(appId, authenticationContext.getPrincipal(), Action.ADMIN);
+    // check that app exists
+    ApplicationSpecification currentSpec = store.getApplication(appId);
+    if (currentSpec == null) {
+      LOG.warn("Application {} not found for upgrade.", appId);
+      return new ApplicationUpdateDetails(new NotFoundException(appId));
+    }
+    ArtifactId currentArtifact = currentSpec.getArtifactId();
+    NamespaceId currentArtifactNamespace =
+        ArtifactScope.SYSTEM.equals(currentArtifact.getScope()) ? NamespaceId.SYSTEM : appId.getParent();
+
+    // Get the artifact with latest version for upgrade.
+    List<ArtifactSummary> availableArtifacts =
+        artifactRepository.getArtifactSummaries(currentArtifactNamespace, currentArtifact.getName(), Integer.MAX_VALUE,
+                                                ArtifactSortOrder.ASC);
+    if (availableArtifacts.isEmpty()) {
+      // This should not be possible as at least the current used artifact should be there.
+      LOG.error("No artifacts found for artifact {} ", currentArtifact);
+      throw new InternalError("Issues in trying to find application artifact for upgrading app " + appId);
```

Comment:
don't throw InternalError. This is a java language error meant for fatal java issues. You should almost never have to throw an Error in normal code.

Reply:
Makes sense.